src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -87,99 +93,72 @@ class PSyDataNode(Statement):
-        self.use_stmt = "use {0}, only: "\
-            .format(self.add_psydata_class_prefix("psy_data_mod")) + \
-            ", ".join(self.add_psydata_class_prefix(symbol) for symbol in
-                      PSyDataNode.symbols)
-
-        # TODO #1185. Much of what is done in this constructor should either
-        # be done in a create() method or in the transformation that adds
-        # this node.
-
-        # A PSyData node always contains a Schedule
-        sched = self._insert_schedule(children)
-        super(PSyDataNode, self).__init__(ast=ast, children=[sched],
-                                          parent=parent)
-
-        # Get or create a symbol table so we can avoid name clashes
-        # when creating variables
-        if parent and hasattr(self.root, 'symbol_table'):
-            symtab = self.root.symbol_table
-        else:
-            # FIXME: This may not be a good solution
-            symtab = SymbolTable()
-
-        # Ensure that we have a container symbol for the API access
-        fortran_module = self.add_psydata_class_prefix(self.fortran_module)
-        try:
-            csym = symtab.lookup(fortran_module)
-        except KeyError:
-            csym = ContainerSymbol(fortran_module)
-            symtab.add(csym)
-
-        # The type symbol for PSyData variables
-        var_name = self.add_psydata_class_prefix("PSyDataType")
-        try:
-            type_sym = symtab.lookup(var_name)
-            if (not isinstance(type_sym.interface, GlobalInterface) or
-                    type_sym.interface.container_symbol is not csym):
-                raise InternalError(
-                    "Symbol table already contains a symbol named '{0}' but "
-                    "its interface does not refer to the '{1}' container.".
-                    format(var_name, csym.name))
-        except KeyError:
-            type_sym = TypeSymbol(var_name, DeferredType(),
-                                  interface=GlobalInterface(csym))
-            symtab.add(type_sym)
-
-        # Store the name of the PSyData variable that is used for this
-        # PSyDataNode. This allows the variable name to be shown in str
-        # (and also, calling create_name in gen() would result in the name
-        # being changed every time gen() is called).
-        psydata_type = UnknownFortranType(
-            "type({0}), save, target ::".format(type_sym.name))
-        self._var_name = symtab.new_symbol(
-            self._psy_data_symbol_with_prefix, symbol_type=DataSymbol,
-            datatype=psydata_type).name
-
-        # Name of the region. In general at constructor time we might
-        # not have a parent subroutine or any child nodes, so
-        # the name is left empty, unless explicitly provided by the
-        # user. If names are not provided here then the region and
-        # module names are set the first time gen() is called (and
-        # then remain unchanged).
+        self._region_identifier = ("", "")
```

Comment:
At first I thought you were right so I removed _region_identifier. However, that broke a test so then I read the comment properly and realised we cannot remove _region_identifier so I've put it back. My understanding is that, if no explicit region name is provided by the user, we generate a region name. We need access to this while generating the code and potentially while generating driver code (for the case of kernel extraction). However, we *don't* want this generated name to be used if a `gen` is performed a second time as the tree might have changed and there may now be other PSyData nodes, meaning the generated name must be changed. Possibly it would be better to cope with this situation by having a separate attribute that records whether the name of the node was originally specified by the user or auto-generated. However, I think (hope!) that's something for another PR and possibly a discussion with @hiker as there may be subtleties I've missed.